### DISCUSSION TITLE
NASM X86/X64 game engine development

### DISCUSSION THREAD

**Discussion by emullallphotography** [community/community#74664] [Programming Help]
### Body

Hi,

I wanted to create a game engine for my game and have it support older hardware by using NASM X86/X64. 

Also,

what are the downsides to using only NASM X86/X64 when it comes to newer hardware?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by arienshibani** [comment]
Creating a game engine using NASM X86/X64 assembly language is quite an ambitious task. Assembly language is very low-level, meaning it's closer to machine code than to human language. This gives you a lot of control over the hardware, which can be beneficial for supporting older hardware. However, it also makes the development process more complex and time-consuming.

As for the downsides of using only NASM X86/X64 for newer hardware, here are a few points to consider:

1. **Performance**: Modern hardware often has features that aren't easily accessible or optimizable with X86/X64 assembly. High-level languages and compilers can often take better advantage of these features.

2. **Portability**: NASM X86/X64 is specific to Intel-based architectures. If you want your game engine to support other architectures (like ARM, which is used in many mobile devices), you'll need to write separate assembly code for each.

3. **Development Time**: Writing in assembly can be very time-consuming. It's much more verbose than high-level languages, and tasks that can be done with a single line of code in a high-level language often require multiple lines in assembly.

4. **Maintenance and Debugging**: Assembly code can be difficult to read and understand, which makes maintaining your code and debugging issues more challenging.

5. **Lack of Modern Features**: Assembly lacks many features of modern high-level languages, such as object-oriented programming, automatic memory management, and extensive standard libraries. This means you'll have to implement a lot of functionality from scratch.

Given these challenges, it's common to write the majority of a game engine in a high-level language like C++ or C#, and only use assembly for performance-critical sections of code where the benefits outweigh the downsides.
